<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-анализу-данных" data-toc-modified-id="Вывод-по-анализу-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод по анализу данных</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Модель-логической-регрессии" data-toc-modified-id="Модель-логической-регрессии-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Модель логической регрессии</a></span></li><li><span><a href="#Модель-дерева-решений" data-toc-modified-id="Модель-дерева-решений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Модель дерева решений</a></span></li><li><span><a href="#Модель-случайного-леса" data-toc-modified-id="Модель-случайного-леса-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Модель случайного леса</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Модель-логической-регрессии" data-toc-modified-id="Модель-логической-регрессии-3.0.1"><span class="toc-item-num">3.0.1&nbsp;&nbsp;</span>Модель логической регрессии</a></span></li><li><span><a href="#Модель-дерева-решений" data-toc-modified-id="Модель-дерева-решений-3.0.2"><span class="toc-item-num">3.0.2&nbsp;&nbsp;</span>Модель дерева решений</a></span></li><li><span><a href="#Модель-случайного-леса" data-toc-modified-id="Модель-случайного-леса-3.0.3"><span class="toc-item-num">3.0.3&nbsp;&nbsp;</span>Модель случайного леса</a></span></li></ul></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Уменьшение выборки</a></span><ul class="toc-item"><li><span><a href="#Модель-логической-регрессии" data-toc-modified-id="Модель-логической-регрессии-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Модель логической регрессии</a></span></li><li><span><a href="#Модель-дерева-решений" data-toc-modified-id="Модель-дерева-решений-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Модель дерева решений</a></span></li><li><span><a href="#Модель-случайного-леса" data-toc-modified-id="Модель-случайного-леса-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Модель случайного леса</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Финальные-комментарии-ревьюера" data-toc-modified-id="Финальные-комментарии-ревьюера-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Финальные комментарии ревьюера</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

RowNumber — индекс строки в данных

CustomerId — уникальный идентификатор клиента

Surname — фамилия

CreditScore — кредитный рейтинг

Geography — страна проживания

Gender — пол

Age — возраст

Tenure — сколько лет человек является клиентом банка

Balance — баланс на счёте

NumOfProducts — количество продуктов банка, используемых клиентом

HasCrCard — наличие кредитной карты

IsActiveMember — активность клиента

EstimatedSalary — предполагаемая зарплата

Целевой признак

Exited — факт ухода клиента

## Подготовка данных

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
from matplotlib import pyplot as plt
from sklearn.utils import shuffle
from sklearn.metrics import r2_score

In [2]:
try:
    data = pd.read_csv('/datasets/Churn.csv')
except FileNotFoundError:
    data = pd.read_csv(r'C:\Users\ASUS\Desktop\Практикум\Data\Churn.csv')
    

In [3]:
data.shape

(10000, 14)

In [4]:
data.describe()

RowNumber    CustomerId   CreditScore           Age       Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  9091.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800     4.997690   
std     2886.89568  7.193619e+04     96.653299     10.487806     2.894723   
min        1.00000  1.556570e+07    350.000000     18.000000     0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000     2.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000     5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000     7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000    10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

In [5]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [6]:
data.sample(10)

RowNumber  CustomerId     Surname  CreditScore Geography  Gender  Age  \
9357       9358    15814405  Chesnokova          418    France  Female   46   
6989       6990    15811877        Shao          700    France  Female   36   
9586       9587    15786418        Chiu          546    France  Female   20   
2606       2607    15715707       Light          657    France    Male   32   
775         776    15752137     McElroy          648    France    Male   33   
3788       3789    15707110    Endrizzi          660   Germany    Male   28   
8010       8011    15787756   Nkemdirim          467   Germany    Male   51   
6219       6220    15662574       Brady          636     Spain    Male   37   
6805       6806    15577988     Skinner          614    France  Female   35   
7899       7900    15683560       Gallo          642    France  Female   40   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
9357     9.0       0.00              1          1               1   
6989     4.0       0.00              2          1               0   
9586     6.0       0.00              1          0               1   
2606     3.0  118829.03              2          1               1   
775      7.0  134944.00              1          1               1   
3788     2.0  170890.05              2          1               0   
8010    10.0  114514.71              2          1               0   
6219     1.0  115137.26              1          1               0   
6805     1.0       0.00              2          1               1   
7899     7.0       0.00              2          1               0   

      EstimatedSalary  Exited  
9357         81014.50       1  
6989        130789.15       0  
9586         20508.85       0  
2606         73127.61       0  
775         117036.38       0  
3788         41758.90       0  
8010        177784.68       1  
6219         52484.01       0  
6805          3342.62       0  
7899        183963.34       0

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


Приведем названия колонок к однуму формату

In [8]:
data.columns = data.columns.str.lower()
data = data.rename(columns={
    'rownumber':'row_number',            
    'customerid':'customer_id',                     
    'creditscore':'credit_score',                                                                        
    'numofproducts':'num_of_products',        
    'hascrcard':'has_cr_card',            
    'isactivemember':'is_active_member',       
    'estimatedsalary':'estimated_salary',    
})

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   row_number        10000 non-null  int64  
 1   customer_id       10000 non-null  int64  
 2   surname           10000 non-null  object 
 3   credit_score      10000 non-null  int64  
 4   geography         10000 non-null  object 
 5   gender            10000 non-null  object 
 6   age               10000 non-null  int64  
 7   tenure            9091 non-null   float64
 8   balance           10000 non-null  float64
 9   num_of_products   10000 non-null  int64  
 10  has_cr_card       10000 non-null  int64  
 11  is_active_member  10000 non-null  int64  
 12  estimated_salary  10000 non-null  float64
 13  exited            10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


Проверим данные на пропуски и дубликаты.

In [10]:
data.duplicated().sum()

0

In [11]:
data.isna().sum()

row_number            0
customer_id           0
surname               0
credit_score          0
geography             0
gender                0
age                   0
tenure              909
balance               0
num_of_products       0
has_cr_card           0
is_active_member      0
estimated_salary      0
exited                0
dtype: int64

Дубликатов нет, а пропуски заполним медианным значением, так как ~10% это много данных 

In [12]:
data['tenure'].fillna(data['tenure'].median(), inplace=True)

In [13]:
data[data['balance']==0]

row_number  customer_id      surname  credit_score geography  gender  \
0              1     15634602     Hargrave           619    France  Female   
3              4     15701354         Boni           699    France  Female   
6              7     15592531     Bartlett           822    France    Male   
11            12     15737173      Andrews           497     Spain    Male   
12            13     15632264          Kay           476    France  Female   
...          ...          ...          ...           ...       ...     ...   
9989        9990     15605622     McMillan           841     Spain    Male   
9992        9993     15657105  Chukwualuka           726     Spain    Male   
9994        9995     15719294         Wood           800    France  Female   
9995        9996     15606229     Obijiaku           771    France    Male   
9997        9998     15584532          Liu           709    France  Female   

      age  tenure  balance  num_of_products  has_cr_card  is_active_member  \
0      42     2.0      0.0                1            1                 1   
3      39     1.0      0.0                2            0                 0   
6      50     7.0      0.0                2            1                 1   
11     24     3.0      0.0                2            1                 0   
12     34    10.0      0.0                2            1                 0   
...   ...     ...      ...              ...          ...               ...   
9989   28     4.0      0.0                2            1                 1   
9992   36     2.0      0.0                1            1                 0   
9994   29     2.0      0.0                2            0                 0   
9995   39     5.0      0.0                2            1                 0   
9997   36     7.0      0.0                1            0                 1   

      estimated_salary  exited  
0            101348.88       1  
3             93826.63       0  
6             10062.80       0  
11            76390.01       0  
12            26260.98       0  
...                ...     ...  
9989         179436.60       0  
9992         195192.40       0  
9994         167773.55       0  
9995          96270.64       0  
9997          42085.58       1  

[3617 rows x 14 columns]

3617 строк с 0 балансом, странно, моё предположение что это отрицательный баланс, пока оставим так. В идеале надо бы уточнить что означают 0 в данных по балансу.

In [14]:
data = data.drop(['row_number', 'customer_id', 'surname'], axis=1)

Избавимся от столбца 'row_number' так как он повторяет индекс таблицы. Так же столбцы 'customer_id' и 'surname' не несут важной информации для нашего исследования, поэтому мы избавимся и от них

In [15]:
data_ohe = pd.get_dummies(data, drop_first=True)

In [16]:
features = data_ohe.drop('exited',axis = 1)
target = data_ohe['exited']

In [17]:
features_train, features_temp, target_train, target_temp = train_test_split(features, target, test_size=0.4, random_state=12345)
features_valid, features_test, target_valid, target_test = train_test_split(features_temp, target_temp, test_size=0.5, random_state=12345)


In [18]:
print('features_train:', features_train.shape)
print('features_valid:', features_valid.shape)
print('features_test: ', features_test.shape)

features_train: (6000, 11)
features_valid: (2000, 11)
features_test:  (2000, 11)


In [19]:
numeric = ['credit_score', 'age', 'balance', 'estimated_salary', 'tenure', 'num_of_products']
scaler = StandardScaler()
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])

### Вывод по анализу данных

* Общее количество строк в таблице 10000 и 14 столбцов.
* Значения имеют разный тип данных int и float, object.Изменены названия 7 столбцов.Дубликатов не обнаружено.
* Убрано 3 столбца из датасета
* В столбце tenor пропуски заменены медианным значением.
* Данные разделены на обучающую, валидационную и тестовую выборки в соотношении 3:1:1

## Исследование задачи

In [20]:
target_train.value_counts()


0    4804
1    1196
Name: exited, dtype: int64

### Модель логической регрессии

In [21]:
%%time
    model_regression = LogisticRegression(random_state=12345)
    model_regression.fit(features_train, target_train)
    predictions_valid = model_regression.predict(features_valid)
    predictions_train = model_regression.predict(features_train)
f1_valid = f1_score(target_valid, predictions_valid)
f1_train = f1_score(target_train, predictions_train)
print("F1 наилучшей модели на валидационной выборке:", f1_valid)
print("F1 наилучшей модели на обучающей выборке:", f1_train)

F1 наилучшей модели на валидационной выборке: 0.33108108108108103
F1 наилучшей модели на обучающей выборке: 0.34378769601930037
CPU times: total: 15.6 ms
Wall time: 15.1 ms


In [22]:
probabilities_valid = model_regression.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
fpr, tpr, thresholds = roc_curve(target_valid, probabilities_one_valid)
print('AUC-ROC:', roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC: 0.7588026784580114


### Модель дерева решений

In [23]:
%%time
model_tree = DecisionTreeClassifier(random_state=12345)

max_depth = range(1, 21)

best_score = 0
best_depth = 0

for depth in max_depth:
    model_tree = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model_tree.fit(features_train, target_train)
    predictions_valid = model_tree.predict(features_valid)
    f1 = f1_score(target_valid, predictions_valid)
    if f1 > best_score:
        best_score = f1
        best_depth = depth

print("Лучшее значение F1 на валидационной выборке:", best_score)
print("Лучшая глубина дерева:", best_depth)

Лучшее значение F1 на валидационной выборке: 0.5696969696969697
Лучшая глубина дерева: 6
CPU times: total: 328 ms
Wall time: 366 ms


In [24]:
model_tree = DecisionTreeClassifier(max_depth=best_depth, random_state=12345)
model_tree.fit(features_train, target_train)
probabilities_valid = model_tree.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
print('AUC-ROC:',roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC: 0.8164631712023421


### Модель случайного леса

In [25]:
%%time
model_forest = RandomForestClassifier(random_state=12345)

n_estimators = range(10, 121, 10)
max_depth = range(1, 21)

best_score = 0
best_n_estimators = 0
best_max_depth = 0

for n in n_estimators:
    for depth in max_depth:
        model_forest = RandomForestClassifier(n_estimators=n, max_depth=depth, random_state=12345)
        model_forest.fit(features_train, target_train)
        predictions_valid = model_forest.predict(features_valid)
        f1 = f1_score(target_valid, predictions_valid)
        if f1 > best_score:
            best_score = f1
            best_n_estimators = n
            best_max_depth = depth

print("Лучшее значение F1 на валидационной выборке:", best_score)
print("Лучшее количество деревьев:", best_n_estimators)
print("Лучшая глубина деревьев:", best_max_depth)

Лучшее значение F1 на валидационной выборке: 0.5913818722139673
Лучшее количество деревьев: 60
Лучшая глубина деревьев: 19
CPU times: total: 1min 18s
Wall time: 1min 20s


In [26]:
model_forest = RandomForestClassifier(n_estimators=best_n_estimators, max_depth=best_max_depth, random_state=12345)
model_forest.fit(features_train, target_train)
probabilities_valid = model_forest.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
print('AUC-ROC:',roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC: 0.8343611442120991


### Вывод

Лучший результат показала модель случайного леса показав:
F1 = 0.5913818722139673 
AUC-ROC = 0.8343611442120991
При количестве деревьев 60 и глубиной 19

## Борьба с дисбалансом

# Увеличение выборки

In [27]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

features_upsampled, target_upsampled = upsample(features_train, target_train, 4)


#### Модель логической регрессии

In [28]:
%%time
balanced_model_regression = LogisticRegression(random_state=12345)
balanced_model_regression.fit(features_upsampled, target_upsampled)
predictions_valid = balanced_model_regression.predict(features_valid)
f1_valid = f1_score(target_valid, predictions_valid)
print("F1 наилучшей модели на валидационной выборке:", f1_valid)

F1 наилучшей модели на валидационной выборке: 0.4888507718696398
CPU times: total: 0 ns
Wall time: 12.6 ms


In [29]:
probabilities_valid = balanced_model_regression.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
fpr, tpr, thresholds = roc_curve(target_valid, probabilities_one_valid)
print('AUC-ROC:',roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC: 0.7636206364664678


#### Модель дерева решений

In [30]:
%%time
balanced_model_tree = DecisionTreeClassifier(random_state=12345)

max_depth = range(1, 21)

best_score = 0
best_depth = 0

for depth in max_depth:
    balanced_model_tree = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    balanced_model_tree.fit(features_upsampled, target_upsampled)
    predictions_valid = balanced_model_tree.predict(features_valid)
    f1 = f1_score(target_valid, predictions_valid)
    if f1 > best_score:
        best_score = f1
        best_depth = depth        
        
print("Лучшее значение F1 на валидационной выборке:", best_score)
print("Лучшая глубина дерева:", best_depth)


Лучшее значение F1 на валидационной выборке: 0.5963791267305644
Лучшая глубина дерева: 5
CPU times: total: 469 ms
Wall time: 481 ms


In [31]:
balanced_model_tree = DecisionTreeClassifier(max_depth=best_max_depth, random_state=12345)
balanced_model_tree.fit(features_upsampled, target_upsampled)
probabilities_valid = balanced_model_tree.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
print('AUC-ROC:',roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC: 0.6760203001469886


#### Модель случайного леса

In [32]:
%%time
balanced_model_forest = RandomForestClassifier(random_state=12345)

n_estimators = range(10, 121, 10)
max_depth = range(1, 21)

best_score = 0
best_n_estimators = 0
best_max_depth = 0

for n in n_estimators:
    for depth in max_depth:
        balanced_model_forest = RandomForestClassifier(n_estimators=n, max_depth=depth, random_state=12345)
        balanced_model_forest.fit(features_upsampled, target_upsampled)
        predictions_valid = balanced_model_forest.predict(features_valid)
        f1 = f1_score(target_valid, predictions_valid)
        if f1 > best_score:
            best_score = f1
            best_n_estimators = n
            best_max_depth = depth

print("Лучшее значение F1 на валидационной выборке:", best_score)
print("Лучшее количество деревьев:", best_n_estimators)
print("Лучшая глубина деревьев:", best_max_depth)

Лучшее значение F1 на валидационной выборке: 0.6291179596174282
Лучшее количество деревьев: 70
Лучшая глубина деревьев: 9
CPU times: total: 1min 46s
Wall time: 1min 47s


In [33]:
balanced_model_forest = RandomForestClassifier(n_estimators=best_n_estimators, max_depth=best_max_depth, random_state=12345)
balanced_model_forest.fit(features_upsampled, target_upsampled)
probabilities_valid = balanced_model_forest.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
print('AUC-ROC:',roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC: 0.8526303691650686


### Уменьшение выборки

In [34]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_zeros_downsampled = features_zeros.sample(frac=fraction, random_state=12345)
    features_downsampled = pd.concat([features_zeros_downsampled] + [features_ones])
    target_downsampled = pd.concat([target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])

    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)

    return features_downsampled, target_downsampled
features_downsampled, target_downsampled = downsample(features_train, target_train, 0.25)

#### Модель логической регрессии

In [35]:
%%time
downsampled_model_regression = LogisticRegression(random_state=12345)
downsampled_model_regression.fit(features_downsampled, target_downsampled)
predictions_valid = downsampled_model_regression.predict(features_valid)
f1_valid = f1_score(target_valid, predictions_valid)
print("F1 наилучшей модели на валидационной выборке:", f1_valid)

F1 наилучшей модели на валидационной выборке: 0.48406546080964685
CPU times: total: 0 ns
Wall time: 9.05 ms


In [36]:
probabilities_valid = downsampled_model_regression.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
fpr, tpr, thresholds = roc_curve(target_valid, probabilities_one_valid)
print('AUC-ROC:',roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC: 0.7622203134545938


#### Модель дерева решений

In [37]:
%%time
downsampled_model_tree = DecisionTreeClassifier(random_state=12345)

max_depth = range(1, 21)

best_score = 0
best_depth = 0

for depth in max_depth:
    downsampled_model_tree = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    downsampled_model_tree.fit(features_downsampled, target_downsampled)
    predictions_valid = downsampled_model_tree.predict(features_valid)
    f1 = f1_score(target_valid, predictions_valid)
    if f1 > best_score:
        best_score = f1
        best_depth = depth        
        
print("Лучшее значение F1 на валидационной выборке:", best_score)
print("Лучшая глубина дерева:", best_depth)


Лучшее значение F1 на валидационной выборке: 0.5943204868154158
Лучшая глубина дерева: 5
CPU times: total: 125 ms
Wall time: 184 ms


In [38]:
downsampled_model_tree = DecisionTreeClassifier(max_depth=best_max_depth, random_state=12345)
downsampled_model_tree.fit(features_downsampled, target_downsampled)
probabilities_valid = downsampled_model_tree.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
print('AUC-ROC:',roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC: 0.7618611593343778


#### Модель случайного леса

In [39]:
%%time
downsampled_model_forest = RandomForestClassifier(random_state=12345)

n_estimators = range(10, 121, 10)
max_depth = range(1, 21)

best_score = 0
best_n_estimators = 0
best_max_depth = 0

for n in n_estimators:
    for depth in max_depth:
        downsampled_model_forest = RandomForestClassifier(n_estimators=n, max_depth=depth, random_state=12345)
        downsampled_model_forest.fit(features_downsampled, target_downsampled)
        predictions_valid = downsampled_model_forest.predict(features_valid)
        f1 = f1_score(target_valid, predictions_valid)
        if f1 > best_score:
            best_score = f1
            best_n_estimators = n
            best_max_depth = depth

print("Лучшее значение F1 на валидационной выборке:", best_score)
print("Лучшее количество деревьев:", best_n_estimators)
print("Лучшая глубина деревьев:", best_max_depth)

Лучшее значение F1 на валидационной выборке: 0.6092184368737475
Лучшее количество деревьев: 10
Лучшая глубина деревьев: 4
CPU times: total: 43.5 s
Wall time: 44 s


In [40]:
downsampled_model_forest = RandomForestClassifier(n_estimators=best_n_estimators, max_depth=best_max_depth, random_state=12345)
downsampled_model_forest.fit(features_downsampled, target_downsampled)
probabilities_valid = downsampled_model_forest.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
print('AUC-ROC:',roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC: 0.8354975834598564


### Вывод

Лучший резултат показала модель случайного леса при количестве деревьев 70 и глубиной 9 с увеличенной выборкой показав:
* F1 = 0.6291179596174282
* AUC-ROC = 0.8518621574047749


## Тестирование модели

In [41]:
features_full_train = pd.concat([features_train, features_valid])
target_full_train = pd.concat([target_train, target_valid])
features_upsampled, target_upsampled = upsample(features_full_train, target_full_train, 4)

In [42]:
final_model = RandomForestClassifier(random_state=12345,max_depth = 9,n_estimators=70)
final_model.fit(features_upsampled, target_upsampled)
predictions_test = final_model.predict(features_test)
print('F1:', f1_score(target_test, predictions_test))

probabilities_valid = final_model.predict_proba(features_test)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_test, probabilities_one_valid)

print('AUC-ROC:', auc_roc)

F1: 0.61198738170347
AUC-ROC: 0.8577107984007699


## Вывод

* Были изученны и подготовленны данные, а так же разделены на 3 выборки - обучающую, валидационную и тестовую 
* Были исследованы модели без учета баланса классов. Лучший результат показала модель случайного леса
* Для борьбы с дисбалансом применили уменьшение и увеличение выборки. Также были подобранны гиперпараметы для модели. Модель случайного леса снова показала лучший результат.
* Финальную модель протестировали на увеличенной выборки при глубине 9 и количестве деревьев 70. 
  В результате получили: F1: 0.61 и AUC-ROC: 0.86  
  
